openedx_external_enrollments/external_enrollments/pathstream_external_enrollment.py

```diff
@@ -1,22 +1,180 @@
+UNCOMPLETED = False
+
+
+class PathstreamTaskExecutionError(Exception):
+    """Exception when the proccess of executing PathstreamExternalEnrollment.execute_upload fails."""
 
 
 class PathstreamExternalEnrollment(BaseExternalEnrollment):
     """
     PathstreamExternalEnrollment class.
     """
+    def __init__(self):
+        self.client = None
+        self.S3_BUCKET = settings.OEE_PATHSTREAM_S3_BUCKET
+        self.S3_FILE = settings.OEE_PATHSTREAM_S3_FILE
+
     def __str__(self):
         return 'pathstream'
 
+    def _init_s3(self):
+        """Define the S3 service client.
+
+        At this point, _init_s3 does not establish any AWS connection,
+        nor verify the credentials. That will happen when calling
+        _download_file or _upload_file."""
+        self.client = boto3.client(
+            's3',
+            aws_access_key_id=settings.OOE_PATHSTREAM_S3_ACCESS_KEY,
+            aws_secret_access_key=settings.OOE_PATHSTREAM_S3_SECRET_KEY,
+        )
+
+    def _download_file(self):
+        """Download the file content from an S3 Bucket.
+
+        :return: file's content (bytes)"""
+        try:
+            response = self.client.get_object(
+                Bucket=self.S3_BUCKET,
+                Key=self.S3_FILE,
+            )
+        except ClientError as error:
+            raise PathstreamTaskExecutionError(
+                '_download_file',
+                'Failed to download the content of the file from S3. Reason: {}'.format(str(error))
+            )
+
+        LOG.info('File content [%s] was downloaded from S3 for [%s]', self.S3_FILE, self.__str__())
+
+        return response['Body'].read()
+
+    def _prepare_new_content(self, content):
+        """This method returns the new content ready to be uploaded to S3. It first sorts the input list
+        by date/time and then encode it since this is the way to upload the file content to S3.
+
+        :param: content (list of strings).
+            The elements of content are expected to be the following comma-separated format:
+            "course,email,2021-08-05 17:53:41.492901,false"
+
+        :return: bytes made up of the data passed in content.
+
+        :raise: AttributeError, IndexError or ValueError in case at least one element in content has a different
+        format than expected."""
+        try:
+            content.sort(
+                #  Extract datetime out of every element in content to be the sorting parameter.
+                key=lambda data: datetime.strptime(data.split(',')[2], settings.OOE_PATHSTREAM_S3_DATETIME_FORMAT)
+            )
+        except (AttributeError, IndexError, ValueError) as error:
+            raise PathstreamTaskExecutionError(
+                '_prepare_new_content',
+                'An enrollment data is invalid. Reason: {} \nNew data: {}'.format(str(error), content)
+            )
+
+        # Convert elements of content to bytes.
+        content = map(lambda data: data.encode(), content)
```

Comment:
Why do you need to encode the data? Aren't they supposed to be UTF-8 already? @Jacatove

Reply:
The elements of `content` need to be in their bytes representations to be  joined. Then, the returned value will be appended to the downloaded file content --which comes also in bytes-- to finally upload it. @Squirrel18